### PR TITLE
Refactor parties generator

### LINKED ISSUE #11573: Refactor for parties_generators
1) removes all reference to `ledger`, instead make `account_state_tbl` a required parameter.
2) provide a function for generation of sequence of parties. This is mainly because generation of sequence of parties needs to decide the set of fee_payer accounts in advance and thread the `account_state_tbl` through the multiple calls of `gen_parties_from`.
3) writes a mli files for the interface of parties generators and writes document for those generators.

### PR DESCRIPTION
**Pull request by ghost-not-in-the-shell**
Explain your changes:
This PR did some refactors around parties generator according to #11573

Explain how you tested your changes:
*

Checklist:

- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

* Closes #11573

### CODE REVIEW

**Review by psteckler — APPROVED** [review]

File: src/lib/mina_generators/parties_generators.ml [hunk 1]
```diff
@@ -1269,11 +1253,10 @@ let gen_parties_from ?failure ?(max_other_parties = max_other_parties)
         let%bind party0 =
           (* Signature authorization to start *)
           let authorization = Control.Signature Signature.dummy in
-          let required_balance_change = Currency.Amount.Signed.zero in
-          gen_party_from ~account_ids_seen ~update ?failure ~authorization
-            ~new_account:new_parties ~permissions_auth ~zkapp_account
-            ~available_public_keys ~required_balance_change ~ledger
-            ~account_state_tbl ?protocol_state_view ?vk ()
+          gen_party_from ~zkapp_account_ids ~account_ids_seen ~update ?failure
+            ~authorization ~new_account:new_parties ~permissions_auth
```
Comment (deepthiskumar) [inline src/lib/mina_generators/parties_generators.ml:1257]: rename `new_parties` to something that says it is a boolean? maybe `is_new_party`?

File: src/lib/mina_generators/parties_generators.ml [hunk 2]
```diff
@@ -1343,10 +1326,10 @@ let gen_parties_from ?failure ?(max_other_parties = max_other_parties)
           in
           (* if we use this account again, it will have a Signature authorization *)
           let permissions_auth = Control.Tag.Signature in
```
Comment (deepthiskumar) [inline src/lib/mina_generators/parties_generators.ml:1328]: So here, we are resetting the permissions to `Signature` (in party0 we set it to whatever was generated). is this because we are generating an additional signed party in each iteration? if so, could you please add a comment that describes this reason?

File: src/lib/mina_generators/parties_generators.ml [hunk 3]
```diff
@@ -1374,30 +1368,73 @@ let gen_parties_from ?failure ?(max_other_parties = max_other_parties)
...
+        ( { account with
+            balance =
+              Currency.Balance.add_signed_amount_flagged account.balance
+                balance_change
+              |> fst
+          }
+        , role ) ) ;
+  (* modify the account balance && nonce precondition of the balancing account to reflect the change*)
+  let other_parties =
+    balancing_party
```
Comment (deepthiskumar) [inline src/lib/mina_generators/parties_generators.ml:1396]: see above comment about using the fee payer account

File: src/lib/mina_generators/parties_generators.ml [hunk 4]
```diff
@@ -745,92 +731,58 @@ let gen_party_body_components (type a b c d) ?(update = None) ?account_id
...
+                  "gen_party_body: all zkapp accounts were new accounts or \
+                   used as fee_payer accounts"
+            | Some (acct, `Ordinary_participant) ->
+                acct
+          else
+            let accts =
+              Account_id.Table.filteri account_state_tbl
+                ~f:(fun ~key:_ ~data:(_, role) ->
+                  match role with
+                  | `Fee_payer | `New_account ->
```
Comment (deepthiskumar) [inline src/lib/mina_generators/parties_generators.ml:771]: In the case of non-zkapp accounts, we could reuse newly created accounts, right?

File: src/lib/mina_generators/parties_generators.ml [hunk 5]
```diff
@@ -244,40 +242,44 @@ let gen_fee (account : Account.t) =
...
+      Balance.(effective_balance < of_formatted_string "1.0") && not new_account
+    then failwith "account has low balance"
     else Balance.of_formatted_string "0.000001"
   in
   let%map (magnitude : Currency.Amount.t) =
-    Currency.Amount.gen_incl Currency.Amount.zero
-      (Currency.Balance.to_amount small_balance_change)
+    if new_account then
+      Currency.Amount.gen_incl
+        (Currency.Amount.of_formatted_string "2.0")
```
Comment (deepthiskumar) [inline src/lib/mina_generators/parties_generators.ml:275]: If we reused new accounts in the same transaction then you might want to increase the initial balance here

File: src/lib/network_pool/transaction_pool.ml [hunk 6]
```diff
@@ -1387,7 +1387,18 @@ struct
                   in
                   Or_error.errorf "Diff failed with verification failure(s): %s"
                     errs_string
-              | Error _ | Ok () ->
+              | Error errs ->
```
Comment (deepthiskumar) [inline src/lib/network_pool/transaction_pool.ml:1390]: revert this change?

**Review by deepthiskumar — CHANGES_REQUESTED** [review]

File: src/lib/mina_generators/parties_generators.ml [hunk 7]
```diff
@@ -1173,9 +1130,26 @@ let gen_parties_from ?failure ?(max_other_parties = max_other_parties)
...
+        | `Fee_payer | `New_account ->
+            false
+        | `Ordinary_participant ->
+            Option.is_some a.zkapp )
+    |> Account_id.Table.keys
+  in
+  Hash_set.add account_ids_seen fee_payer_acct_id ;
+  let%bind balancing_party =
+    let authorization = Control.Signature Signature.dummy in
+    gen_party_from ?failure ~permissions_auth:Control.Tag.Signature
```
[the remark this answers is not in the document]
Comment (ghost-not-in-the-shell) [reply src/lib/mina_generators/parties_generators.ml:1147]: I was doing this because I don't know whether random generated party would increment the nonce or not. Since balancing party would be the head of the list of parties, so I pre-generated it in order to prepare other parties for the change  of nonce.

File: src/lib/mina_generators/parties_generators.ml [hunk 8]
```diff
@@ -1256,7 +1230,17 @@ let gen_parties_from ?failure ?(max_other_parties = max_other_parties)
...
-              let%map tag = Control.Tag.gen in
+              (* Since zkapp account would trigger new account generation, and for new account
+                 verification keys are not in the ledger, so for now we can only use `Signature`
+                 or `None_given` for authorization. This could be fixed by passing the list
+                 of zkapp account into the generator function
+              *)
+              let%map tag =
+                if new_parties then
+                  Quickcheck.Generator.of_list
+                    [ Control.Tag.Signature; None_given ]
```
[the remark this answers is not in the document]
Comment (ghost-not-in-the-shell) [reply src/lib/mina_generators/parties_generators.ml:1241]: Ah. I don't know that. Let me fix it.

[the remark this answers is not in the document]
Comment (ghost-not-in-the-shell) [reply src/lib/mina_generators/parties_generators.ml:1241]: But the auth would be changed back to signature in the next party. I think it should be ok.

File: src/lib/mina_generators/parties_generators.ml (hunk 1 above)
Comment (ghost-not-in-the-shell) [reply src/lib/mina_generators/parties_generators.ml:1257]: The `new_parties` variable would be replaced by another variable in my token accounts PR. So I think it's ok to keep it this way now to reduce merge conflict between my 2 PRs.

File: src/lib/mina_generators/parties_generators.ml (hunk 2 above)
Comment (ghost-not-in-the-shell) [reply src/lib/mina_generators/parties_generators.ml:1328]: The original logic of this `gen_parties_with_dynamic_logic` function comes with this logic. I guess it's because it's easier to implement because we could always assume every account comes with signature authorization. There's a comment describe this behavior but there's no comment for the reason behind this.

File: src/lib/mina_generators/parties_generators.ml (hunk 4 above)
Comment (ghost-not-in-the-shell) [reply src/lib/mina_generators/parties_generators.ml:771]: If we only stick to `signature` and `non-given` authorization.

File: src/lib/mina_generators/parties_generators.ml (hunk 5 above)
Comment (ghost-not-in-the-shell) [reply src/lib/mina_generators/parties_generators.ml:275]: I gave it a small balance because the balance change in party is pretty small. I can make it bigger.

File: src/lib/mina_generators/parties_generators.ml [hunk 9]
```diff
@@ -16,9 +16,7 @@ type failure =
       | `Token_symbol
       | `Balance ]
 
-(* whether we should create a new account in the ledger *)
-let should_create_account ?account_id ~new_account ~zkapp_account =
-  new_account || (zkapp_account && Option.is_none account_id)
+type role = [ `Fee_payer | `New_account | `Ordinary_participant ]
```
[the remark this answers is not in the document]
Comment (ghost-not-in-the-shell) [reply src/lib/mina_generators/parties_generators.ml:19]: This would also be used in `account_state_tbl`. It would be easier to use if it's just a polymorphic variant (no imports and dependencies).

**Review by lk86 — APPROVED** [review]
Product ✅

File: src/lib/network_pool/transaction_pool.ml (hunk 6 above)
Comment (ghost-not-in-the-shell) [reply src/lib/network_pool/transaction_pool.ml]: This change has been reverted in the previous PR.

**Review by deepthiskumar — APPROVED** [review]
